sound/soc/intel/boards/sof_pcm512x.c

```diff
@@ -0,0 +1,474 @@
+	struct list_head head;
+	struct snd_soc_dai *codec_dai;
+	int device;
+};
+
+struct sof_card_private {
+	struct list_head hdmi_pcm_list;
+	bool common_hdmi_codec_drv;
+};
+
+static int sof_pcm512x_quirk_cb(const struct dmi_system_id *id)
+{
+	sof_pcm512x_quirk = (unsigned long)id->driver_data;
+	return 1;
+}
+
+static const struct dmi_system_id sof_pcm512x_quirk_table[] = {
+	{
+		.callback = sof_pcm512x_quirk_cb,
+		.matches = {
+			DMI_MATCH(DMI_SYS_VENDOR, "AAEON"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "UP-CHT01"),
+		},
+		.driver_data = (void *)(SOF_PCM512X_SSP_CODEC(2)),
+	},
+	{}
+};
+
+static int sof_hdmi_init(struct snd_soc_pcm_runtime *rtd)
+{
+	struct sof_card_private *ctx = snd_soc_card_get_drvdata(rtd->card);
+	struct snd_soc_dai *dai = rtd->codec_dai;
+	struct sof_hdmi_pcm *pcm;
+
+	pcm = devm_kzalloc(rtd->card->dev, sizeof(*pcm), GFP_KERNEL);
+	if (!pcm)
+		return -ENOMEM;
+
+	/* dai_link id is 1:1 mapped to the PCM device */
+	pcm->device = rtd->dai_link->id;
+	pcm->codec_dai = dai;
+
+	list_add_tail(&pcm->head, &ctx->hdmi_pcm_list);
+
+	return 0;
+}
+
+static int sof_pcm512x_codec_init(struct snd_soc_pcm_runtime *rtd)
+{
+	struct snd_soc_component *codec = rtd->codec_dai->component;
+
+	snd_soc_component_update_bits(codec, PCM512x_GPIO_EN, 0x08, 0x08);
+	snd_soc_component_update_bits(codec, PCM512x_GPIO_OUTPUT_4, 0x0f, 0x02);
+	snd_soc_component_update_bits(codec, PCM512x_GPIO_CONTROL_1,
+				      0x08, 0x08);
+
+	return 0;
+}
+
+static int aif1_startup(struct snd_pcm_substream *substream)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_component *codec = rtd->codec_dai->component;
+
+	snd_soc_component_update_bits(codec, PCM512x_GPIO_CONTROL_1,
+				      0x08, 0x08);
+
+	return snd_pcm_hw_constraint_single(substream->runtime,
+			SNDRV_PCM_HW_PARAM_RATE, 48000);
+}
+
+static void aif1_shutdown(struct snd_pcm_substream *substream)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_component *codec = rtd->codec_dai->component;
+
+	snd_soc_component_update_bits(codec, PCM512x_GPIO_CONTROL_1,
+				      0x08, 0x00);
+}
+
+static const struct snd_soc_ops sof_pcm512x_ops = {
+	.startup = aif1_startup,
+	.shutdown = aif1_shutdown,
+};
+
+static struct snd_soc_dai_link_component platform_component[] = {
+	{
+		/* name might be overridden during probe */
+		.name = "0000:00:1f.3"
+	}
+};
+
+#if IS_ENABLED(CONFIG_SND_HDA_CODEC_HDMI) || \
+	IS_ENABLED(CONFIG_SND_SOC_HDAC_HDMI)
+static struct snd_soc_jack sof_hdmi[3];
+
+static int sof_card_late_probe(struct snd_soc_card *card)
+{
+	struct sof_card_private *ctx = snd_soc_card_get_drvdata(card);
+	struct snd_soc_component *component = NULL;
+	char jack_name[NAME_SIZE];
+	struct sof_hdmi_pcm *pcm;
+	int err;
+	int i = 0;
+
+	/* HDMI is not supported by SOF on Baytrail/CherryTrail */
+	if (is_legacy_cpu)
+		return 0;
+
+	if (list_empty(&ctx->hdmi_pcm_list))
+		return -EINVAL;
+
+	if (ctx->common_hdmi_codec_drv) {
```

Comment:
To future-proof, we should handle  (!ctx->common_hdmi_codec_drv)  case also such that we "return 0" i.e. HDMI is not supported rather than reporting an error. Applies to the hdmi_pcm_list check as well.

Reply:
@kv2019i do we need to check `mach->mach_params.common_hdmi_codec_drv` at all? I thought if we drop HDAC_HDMI, we ignore that flag completely? Or should we check and if it *isn't* set, skip HDMI registration and return 0?

Reply:
@lyakh We still need to check it as long as we support it for some platforms. Basicly only case it would be false is a kernel configuration error (machine driver should bring in right modules via kconfig), but it's still needs to be checked.

Reply:
Not following @kv2019i, we don't select HDAC_HDMI here and don't make references to hdac_hdmi provided symbols, when could this be problematic (I am only talking about this specific machine driver)

Reply:
@plbossart The user could forcibly disable common_hdmi_codec mode (via kernel module parameter). It's an invalid config for sure, but we should not at least crash in this case. hda_dsp_hdmi_build_controls() expects the codec driver to be loaded, which it wouldn't be in this case.

Reply:
@kv2019i ok, I forgot about the kernel parameter. Do we actually still have a reason to keep it? It generates more problems than solutions I am afraid.

Reply:
@plbossart We still have some issues with the common driver on Chrome devices, so I'm afraid we have to keep it at least for a few stable cycles. I'm afraid. But yeah, supporting both has generated a lot of work. If/when we get the new UCM2 to longterm distro releases, doing this kind of changes will be easier in the future.

Reply:
ok, but this specific driver is mostly for tests, so I suggest we don't bother with the 'uncommmon' hdmi driver.

Reply:
@plbossart I'm trying to prepare a PR for you, so I understand I'll remove `mach->mach_params.common_hdmi_codec_drv` completely